### PR TITLE
[FLINK-28048][connectors] Introduce Source API alternative to FiniteTestSource

### PR DESCRIPTION
**Pull request by afedulov**
## What is the purpose of the change

FiniteTestSource used for testing various bulk writers is based on the deprecates `SourceFunction` API. This PR introduces an alternative that provides identical functionality with a FLIP-27-based Source.

## Verifying this change
This PR introduces changes to tests utilities hence the verification is implicit.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no) 
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    // ------------------------------------------------------------------------
+
+    /**
+     * Verifies that all elements in the iterable are non-null, and are of the given class, or a
+     * subclass thereof.
+     *
+     * @param elements The iterable to check.
+     * @param viewedAs The class to which the elements must be assignable to.
+     */
+    public void checkIterable(Iterable<OUT> elements, Class<?> viewedAs) {
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java:166]: Does this need to be public?

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    private static final Logger LOG = LoggerFactory.getLogger(IndexLookupGeneratorFunction.class);
+
+    /** The (de)serializer to be used for the data elements. */
+    private final TypeSerializer<OUT> serializer;
+
+    /** The actual data elements, in serialized form. */
+    private byte[] elementsSerialized;
+
+    private int numElements;
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java:68]: Can this be final? We only seem to assign this on object creation.

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+    @VisibleForTesting
+    @Nullable
+    public TypeSerializer<OUT> getSerializer() {
+        return serializer;
+    }
+
+    private void serializeElements(Iterable<OUT> elements) throws IOException {
+        Preconditions.checkState(serializer != null, "serializer not set");
+        LOG.info("Serializing elements using  " + serializer);
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java:94]:
```suggestion
        LOG.info("Serializing elements using {}", serializer);
```

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+ * A stream generator function that returns elements from the collection based on their index.
+ *
+ * <p>This generator function serializes the elements using Flink's type information. That way, any
+ * object transport using Java serialization will not be affected by the serializability of the
+ * elements.
+ *
+ * @param <OUT> The type of elements returned by this function.
+ */
+@Internal
+public class IndexLookupGeneratorFunction<OUT> implements GeneratorFunction<Long, OUT> {
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java:56]: I tried to understand the `GeneratorFunction` interface. That interface generates values but in turn requires a generator as input. In that sense it is a MapFunction. Maybe call it that way?

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    private void emitElements(ReaderOutput<O> output) {
+        iterator = currentSplit.getIterator();
+        while (iterator.hasNext()) {
+            E next = iterator.next();
+            O converted = convert(next);
+            output.collect(converted);
+        }
+        // Always wait for two snapshots after emitting the elements
+        snapshotsToWaitFor = 2;
+        snapshotsCompleted = 0;
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java:129]: Why two snapshots?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSourceITCase.java [hunk 6]
```diff
@@ -1074,6 +1078,7 @@ private static void assertResults(CloseableIterator<Row> iterator, List<Row> exp
         List<Row> result = CollectionUtil.iteratorToList(iterator);
         iterator.close();
         result.sort(Comparator.comparingInt(o -> (Integer) o.getField(0)));
+        //        result.forEach((r) -> System.out.println(">>>" + r));
```
Comment (mxm) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveTableSourceITCase.java:1081]: Keep this?

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+@Experimental
+public class SourceReaderWithCheckpointsLatch<
+                E, O, IterT extends Iterator<E>, SplitT extends IteratorSourceSplit<E, IterT>>
+        extends IteratorSourceReaderBase<E, O, IterT, SplitT> {
+
+    private final GeneratorFunction<E, O> generatorFunction;
+
+    private BooleanSupplier couldExit;
+    private int snapshotsCompleted;
+    private int snapshotsToWaitFor = Integer.MAX_VALUE;
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java:58]: Is this default value used?

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    private boolean done;
+
+    public SourceReaderWithCheckpointsLatch(
+            SourceReaderContext context,
+            GeneratorFunction<E, O> generatorFunction,
+            @Nullable BooleanSupplier couldExit) {
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java:64]:
```suggestion
            @Nullable BooleanSupplier safeToExit) {
```

```suggestion
            @Nullable BooleanSupplier canExit) {
```

```suggestion
            @Nullable BooleanSupplier shouldExit) {
```


These make more sense to me.

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+
+        try {
+            for (OUT element : elements) {
+                serializer.serialize(element, wrapper);
+            }
+        } catch (Exception e) {
+            throw new IOException("Serializing the source elements failed: " + e.getMessage(), e);
+        }
+        this.elementsSerialized = baos.toByteArray();
+    }
```
[the remark this answers is not in the document]
Comment (afedulov) [reply flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java:106]: After https://github.com/apache/flink/pull/23553 finally get's merged I will most likely simply extend FromElementsGeneratorFunction - it contains the same functionality.

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java (hunk 5 above)
Comment (afedulov) [reply flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java:129]: The whole endeavor is required to have a drop-in replacement for https://github.com/apache/flink/blob/release-1.18/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/FiniteTestSource.java#L33-L35 without modifying the logic of existing tests that rely on it. This is how FiniteTestSource works.

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java (hunk 8 above)
Comment (afedulov) [reply flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java:64]: Good idea, it is based on the existing naming in FiniteTestSource
https://github.com/apache/flink/blob/0234027c55ce743e3340391c77348f100f712e90/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/FiniteTestSource.java#L77
But changing it makes sense.

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java (hunk 4 above)
Comment (afedulov) [reply flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/functions/IndexLookupGeneratorFunction.java:56]: Do you mean `IndexLookupMapFunction` ? `GeneratorFunction` is a primary interface in the datagen connector and it is meant to work in a specific context of the SourceReader. Naming it a MapFunction would be a bit confusing.

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java (hunk 7 above)
Comment (afedulov) [reply flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java:58]: Good catch! It was in before, but not in the recent implementation.

**Comment by afedulov** [comment]
@mxm thanks! Addressed your comments.

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+ * checkpoints to complete, 3) then re-emits the same elements before 4) waiting for another two
+ * checkpoints and 5) exiting.
+ *
+ * <p>This lockstep execution is possible because {@code pollNext} and {@code snapshotState} are
+ * executed in the same thread and the fact that {@code pollNext} can emit N elements at once. This
+ * reader is meant to be used solely for testing purposes as the substitution for the {@code
+ * FiniteTestSource} which implements the deprecated {@code SourceFunction} API.
+ */
+@Experimental
+public class SourceReaderWithCheckpointsLatch<
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java:50]:
```suggestion
public class DoubleEmittingSourceReaderWithCheckpointsInBetween<
```

Is this more like what it's doing?

**Review by mxm — COMMENTED** [review]
I think the implementation is a lot harder to understand with the new API. I wonder, can we create an adapter for SourceFunction-like implementations based on the Source API? We could then keep all the existing code and just wrap it.

**Comment by afedulov** [comment]
There are some things that can probably be done, but not in the scope of this PR. You can see how much complexity of the new Source API is already hidden behind the `IteratorSourceReaderBase`, the `DataGeneratorSource` and the `GeneratorFunction`. We are getting there, but in order to come up with meaningful abstractions we need to learn more about common patterns and we do so during migration of the existing SourceFunctions. I do not think there is an easy way to some up with something completely generic because of how the new Source imposes implementation of all of the low-level details with very specific classes (Splits, SplitEnumerators, serializers). See, for instance, what is going on in the `NumberSequenceSource` that we wrap around in the `DataGeneratorSource` that this implementation relies on:
https://github.com/apache/flink/blob/e4b000818c15b5b781c4e5262ba83bfc9d65121a/flink-core/src/main/java/org/apache/flink/api/connector/source/lib/NumberSequenceSource.java#L218

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java (hunk 10 above)
Comment (afedulov) [reply flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/SourceReaderWithCheckpointsLatch.java]: My idea was at some point to pass the number of checkpoints from the outside, but sure, we can reflect this number in the name until then. Renamed.

**Comment by afedulov** [comment]
> We could then keep all the existing code and just wrap it.

The execution model is fundamentally different - a while loop that runs directly in the function vs an async loop that runs one level "above" and synchronizes via completable futures instead of exposing an explicit lock. I do not think it is possible to create such an adapter because we do not have access to that higher-level loop.

File: flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/DataGenerators.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+     *
+     * <p>Creates a source that emits provided {@code data}, waits for two checkpoints and emits the
+     * same {@code data } again. It is intended only to be used for test purposes. See {@link
+     * DoubleEmittingSourceReaderWithCheckpointsInBetween} for details.
+     *
+     * @param typeInfo The type information of the elements.
+     * @param data The collection of elements to create the stream from.
+     * @param <OUT> The type of the returned data stream.
+     */
+    public static <OUT> DataGeneratorSource<OUT> fromDataWithSnapshotsLatch(
```
Comment (mxm) [inline flink-connectors/flink-connector-datagen/src/main/java/org/apache/flink/connector/datagen/source/DataGenerators.java:46]:
```suggestion
    public static <OUT> DataGeneratorSource<OUT> doubleEmittingWithCheckpointsInBetween(
```

**Review by mxm — COMMENTED** [review]
I'm just slightly concerned with the complexity here. I'm honestly wondering who would still be able to understand this code, if not very familiar with the new source design AND the historic knowledge of how the old testing source worked.

A general SourceFunction adapter may be too much work / not feasible, but we can at least try to keep things as simple and easy to understand as possible, including the naming.

**Comment by mxm** [comment]
LGTM. I would appreciate just another set of eyes before merging. Do you think you could have a look @zentol?

**Comment by tweise** [comment]
LGTM from my side, based on a quick review after @mxm had already covered the details.

Regarding complexity: FLIP-27 sources are without doubt more complex for the source implementation/author. That was the price to push past the limitations of the (eventually too simplistic) old SourceFunction and accommodate the rather diverse and complex requirements of feature rich source connectors. 

Part of FLIP-27 was to reduce implementation complexity by providing building blocks like abstractions for readers and fetchers. Changes proposed here serve a limited/specific scenario for testing. The goal is not to also modify the consuming tests, which would be risky from a regression angle and perhaps not feasible. While it takes a bit of effort to understand the implementation, I also see no obvious winner from an abstraction perspective, at least not one that would be helpful for general connector needs.

Please also consider that this effort was delayed till now, lacking criticality beyond finally sunsetting SourceFunction. Given that @afedulov has already put significant effort into this, I would personally focus further efforts elsewhere.

**Review by mxm — COMMENTED** [review]
@afedulov Let me know when this is ready. I can merge then.

**Comment by afedulov** [comment]
@flinkbot run azure

**Comment by afedulov** [comment]
@flinkbot run azure

**Review by leonardBang — COMMENTED** [review]
Thanks @afedulov for the contribution, I just merged your another related PR and plz rebase this PR. Before review this PR, I have some concern with @mxm that the cost to correctly understand the complexity without a common abstraction for new Source, it's hard to to developer and even experienced contributor. Maybe it does not directly influence this PR, but I'm not sure we can migrate all SourceFunctions one by one in this PR way.

**Comment by afedulov** [comment]
@leonardBang , thanks for sharing your concerns, I understand where they are coming from. Here are my thoughts:
- This is probably the trickiest source to migrate, since it artificially imposes synchronization requirement between the sources and the sinks. This is not the normal case and I do not expect users or connector implementors to have to deal with these intricate details. The aim of this PR is solely to provide a drop-in replacement for the FiniteTestSource that then can be utilized as is.  This is how the "normal" case looks like: https://github.com/apache/flink/pull/21774/files
- The complexity of the implementation stems from two main aspects:
  - The inherent complexity of the new Source API
  - The fact that the execution model is completely different. The "pull" loop is outside of our reach. In the new Source API we are essentially implementing an async functions that are, by design, supposed to be oblivious of the context it which it is called. There is no checkpoint lock anymore that we can hold to prevent the data from being pulled while the sink finishes flushing data.
- To be honest I feel we are a bit lucky that this is was at all possible to achieve feature parity with the SourceFunction without the the checkpoint lock and without modifying the core of the new Source API itself :) Maybe as we go further we can detect more common patterns to further abstract away, but at the moment I do not have ideas for something meaningful.

**Comment by leonardBang** [comment]
> @leonardBang , thanks for sharing your concerns, I understand where they are coming from. Here are my thoughts:
> 
> * This is probably the trickiest source to migrate, since it artificially imposes synchronization requirement between the sources and the sinks. This is not the normal case and I do not expect users or connector implementors to have to deal with these intricate details. The aim of this PR is solely to provide a drop-in replacement for the FiniteTestSource that then can be utilized as is.
> * This is how the "normal" case would look like: https://github.com/apache/flink/pull/21774/files
> * The complexity of the implementation stems from two main aspects:
>   
>   * The inherent complexity of the new Source API
>   * The fact that the execution model is completely different. The "pull" loop is outside of our reach. In the new Source API we are essentially implementing an async functions that are, by design, supposed to be oblivious of the context it which it is called. There is no checkpoint lock anymore that we can hold to prevent the data from being pulled while the sink finishes flushing data.
> * To be honest I feel we are a bit lucky that this is was at all possible to achieve feature parity with the SourceFunction without the the checkpoint lock and without modifying the core of the new Source API itself :) Maybe as we go further we can detect more common patterns to further abstract away, but at the moment I do not have ideas for something meaningful that could help hide any of this complexity.

Thanks @afedulov for the detailed explanation, the normal case makes sense to me. +1 from my side  base on other committers' +1.

**Review by leonardBang — APPROVED** [review]

**Comment by afedulov** [comment]
@flinkbot run azure

**Comment by afedulov** [comment]
@flinkbot run azure

**Pull request merged**